### PR TITLE
Multisig account feature 

### PR DESCRIPTION
**Pull request by hansl**
This adds a new Module that actually only implements an account feature. Its attribute ID is not advertised.

There are a few holes in the many-rs library that makes this necessary. It works, but it's not ideal. A refactor would be suggested in the future, after pre-alpha.

### CODE REVIEW

File: src/many/src/server/module/_9_account/features/multisig.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    pub approved: bool,
+}
+
+#[derive(Clone, Encode, Decode)]
+#[cbor(map)]
+pub struct InfoReturn {
+    #[n(0)]
+    pub memo: Option<String>,
+
+    #[n(1)]
```
Comment (fmorency) [inline src/many/src/server/module/_9_account/features/multisig.rs:152]: 1 and 2 are swapped vs the spec

File: src/many/src/server/module/_9_account/features/multisig.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    pub submitter: Identity,
+
+    #[n(2)]
+    pub transaction: TransactionInfo,
+
+    #[n(3)]
+    pub approvers: BTreeMap<Identity, ApproverInfo>,
```
Comment (fmorency) [inline src/many/src/server/module/_9_account/features/multisig.rs:159]: You should define `ApproverInfo` in the spec. It's a `bool` in the spec. Here it's a map.

File: src/many/src/server/module/_9_account/features/multisig.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+#[cbor(map)]
+pub struct WithdrawArg {
+    #[n(0)]
+    pub token: ByteVec,
+}
+
+pub type WithdrawReturn = EmptyReturn;
+
+#[many_module(name = AccountMultisigModule, namespace = account, many_crate = crate)]
+pub trait AccountMultisigModuleBackend: Send {
```
Comment (fmorency) [inline src/many/src/server/module/_9_account/features/multisig.rs:206]: `multisigSetDefaults` endpoint missing

**Review by fmorency — COMMENTED** [review]
Thanks! Looking good!

Here are some comments on this Friday 13th. I will probably have more comments following your answers both here and on your spec PR.

I would also want to see some (unit) tests & I will want to play with this feature from `many-framework` as well.

File: src/many/src/types/ledger.rs [hunk 4]
```diff
@@ -434,53 +538,190 @@ impl Encode for TransactionInfo {
...
+                    .encode(revoker)?;
+            }
+            TransactionInfo::MultisigExecute {
+                account,
+                token,
+                executer,
+            } => {
+                e.map(4)?
+                    .u8(0)?
+                    .encode(TransactionKind::MultisigRevoke)?
```
Comment (fmorency) [inline src/many/src/types/ledger.rs:608]: Should be `MultisigExecute`

File: src/many/src/types/ledger.rs [hunk 5]
```diff
@@ -434,53 +538,190 @@ impl Encode for TransactionInfo {
...
+                    .encode(executer)?;
+            }
+            TransactionInfo::MultisigWithdraw {
+                account,
+                token,
+                withdrawer,
+            } => {
+                e.map(4)?
+                    .u8(0)?
+                    .encode(TransactionKind::MultisigRevoke)?
```
Comment (fmorency) [inline src/many/src/types/ledger.rs:623]: Should be `MultisigWithdraw`

File: src/many/src/types/ledger.rs [hunk 6]
```diff
@@ -434,53 +538,190 @@ impl Encode for TransactionInfo {
...
 impl<'b> Decode<'b> for TransactionInfo {
     fn decode(d: &mut Decoder<'b>) -> Result<Self, minicbor::decode::Error> {
-        let _len = d.map()?.ok_or(minicbor::decode::Error::Message(
+        use TransactionInfo::*;
+        let len = d.map()?.ok_or(minicbor::decode::Error::Message(
             "Invalid transaction type.",
         ))?;
 
         if d.u8()? != 0 {
             return Err(minicbor::decode::Error::Message("Invalid TransactionKind"));
```
Comment (fmorency) [inline src/many/src/types/ledger.rs:669]: Should probably be "Missing TransactionKind" or something like that.

File: src/many/src/server/module/_9_account/features/multisig.rs (hunk 2 above)
Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:196]: It's not a bool, it's

```
{
            ; Whether they approved or not.
            0 => bool
        }
```

Doesn't have a type name in the spec.

File: src/many/src/server/module/_9_account/features/multisig.rs (hunk 1 above)
Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:189]: I don't see it. Let's talk later.

Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:189]: Done.

File: src/many/src/server/module/_9_account/features/multisig.rs (hunk 3 above)
Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:243]: Implemented

File: src/many/src/server/module/_9_account/features/multisig.rs [hunk 7]
```diff
@@ -184,13 +188,13 @@ pub struct ApproverInfo {
 #[cbor(map)]
 pub struct InfoReturn {
     #[n(0)]
-    pub memo: Option<ByteVec>,
+    pub memo: Option<String>,
```
Comment (fmorency) [inline src/many/src/server/module/_9_account/features/multisig.rs:191]: Either fix the spec to make it optional or fix it here to make it mandatory

**Review by fmorency — COMMENTED** [review]
Nothing major.

I'm waiting for the `setDefault` clarification in the specification before approving.

File: src/many/src/server/module/_9_account/features/multisig.rs (hunk 7 above)
Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:191]: This is optional.

File: src/many/src/server/module/_9_account/features/multisig.rs [hunk 8]
```diff
@@ -203,8 +207,29 @@ pub struct InfoReturn {
...
     #[n(6)]
     pub timeout: Timestamp,
+
+    #[n(7)]
+    pub data: Option<ByteVec>,
+}
+
+#[derive(Clone, Encode, Decode)]
+#[cbor(map)]
+pub struct SetDefaultsArg {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many/src/server/module/_9_account/features/multisig.rs:217]: Good catch. I've set `Arg` for all structs in this PR. I'll update it all.

**Review by fmorency — APPROVED** [review]
LGTM!

**Pull request merged**